src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,506 @@
+    private int numSlotsForIndexRebuild;
+
+
+    // Each slot in the index array can be in one of 3 states, depending on the value of its second (offset) field :
+    // Offset >= 1 : Filled, the data in the hash field and the index field is valid. Subtracting one from the offset
+    // will yield a valid index into the KV array.
+    // Offset == 0 : Unused, the data in the hash field and the index field is NOT valid, and the slot was never filled
+    // with valid data.
+    // Offset == -1 : Deleted, the data in the hash field and the index field is NOT valid, but the slot used to be
+    // occupied with valid data before.
+    private static final int INDEX_SLOT_UNUSED_MARKER = 0;
+    private static final int INDEX_SLOT_DELETED_MARKER = -1;
+
+    // Number of hash entries, not array positions (in general, capacities and sizes are always in entries)
+    public static final int DEFAULT_INITIAL_CAPACITY = 8;
+    public static final float THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD = 0.7f;
+
+    // temporary debugging aid so that its obvious the class is being used without being spam
+    private static boolean LOG_FIRST_ACCESS = false;
+
+    public CompactHashStore() {
+        this(DEFAULT_INITIAL_CAPACITY);
+    }
+
+    // private non-allocating constructor for .copy(), not part of the interface
+    private CompactHashStore(int[] index, Object[] kvs, int kvStoreInsertionPos, int numSlotsForIndexRebuild) {
+        this.index = index;
+        this.kvStore = kvs;
+        this.kvStoreInsertionPos = kvStoreInsertionPos;
+        this.numSlotsForIndexRebuild = numSlotsForIndexRebuild;
+    }
+
+    public CompactHashStore(int capacity) {
+        // temporary and ugly
+        if (LOG_FIRST_ACCESS) {
+            System.out.println("~~~~~~~~~~~~~~~~~~~~~~~ Using Compact Hashes from Hash Tables That Don't Hate You.");
+            LOG_FIRST_ACCESS = false; // without this we would spam the output stream
+        }
+        assertConstructorPreconditions(capacity);
+
+        int kvCapacity = roundUpwardsToNearestPowerOf2(capacity);
+        // the index array needs to be a little bit sparse for good performance (i.e. low load factor)
+        // so to store 8 key-value entries an index of exactly 8 slots is too crowded
+        int indexCapacity = 2 * kvCapacity;
+
+        // All 0s by default, so all slots are marked empty for free
+        this.index = new int[2 * indexCapacity];
+        this.kvStore = new Object[2 * kvCapacity];
+        this.kvStoreInsertionPos = 0; // always increases by 2, never decreases
+        this.numSlotsForIndexRebuild = (int) (this.index.length / 2 * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD);
+    }
+
+    private static int roundUpwardsToNearestPowerOf2(int num) {
+        return Integer.highestOneBit(num - 1) << 1;
+    }
+
+    private static void assertConstructorPreconditions(int capacity) {
+        if (capacity < 1 || capacity > (1 << 28)) {
```

Comment:
Why 28? I realize we need the number to be effectively half the max array size so we can store pairs of values, but this seems smaller than I'd expect to be supported.

Reply:
Well, that number will be rounded upwards to the nearest power of 2. So 2^28 + 1 wil be rounded upwards to 2^29. That will be the kv capacity, the index capacity is 2 times that, so 2^30. That's a problem, because the capacity is in entries, so we allocate twice 2^30 array entries for the index, so 2^31, but that's negative !

The thing we can remove to mitigate that is the requirement that the kv size be a multiple of 2, there is no real reason for that, we technically only require that index.length is a power of 2 for efficient mod. So we technically can use the supplied capacity as-is for the KV, and then round upwards for the index capacity. 

But there's a pathological edge case there, what if the supplied capacity was already a power of 2 ? rouding "upwards" will yield the same power of 2, and then you end up with a KV and an index array of the same size, which is a too crowded index.

Reply:
OK so 2^28 is correct then, could you add a comment explaining that?